src/main/java/com/terraformers/smolder/api/SmolderBiomeRegistry.java

```
@@ -0,0 +1,252 @@
+				SUBBIOMES.put(parent, subbiomes);
+			}
+			if (OVERRIDE || !subbiomes.contains(subbiome))
+			{
+				Float weight = WEIGHTS.get(parent);
+				if (weight == null)
+					weight = 1F;
+				subbiomes.add(subbiome);
+				weight = subbiome.addWeight(weight);
+				PARENTS.put(subbiome, parent);
+				WEIGHTS.put(parent, weight);
+			}
+		}
+		return subbiome;
+	}
+	
+	/**
+	 * Return a list of all registered biomes.
+	 * @return {@link List} of {@link SmolderBiome}.
+	 */
+	public static List<SmolderBiome> getBiomes()
+	{
+		List<SmolderBiome> biomes = new ArrayList<SmolderBiome>();
+		BIOMES.forEach((biome) -> {
+			if (!biomes.contains(biome))
+				biomes.add(biome);
+		});
+		return ImmutableList.copyOf(biomes);
+	}
+	
+	/**
+	 * Returns an edge for biome or null;
+	 * @param biome - parent {@link SmolderBiome};
+	 * @return {@link SmolderBiome} edge.
+	 */
+	public static SmolderBiome getEdge(SmolderBiome biome)
+	{
+		return EDGES.get(biome);
+	}
+	
+	/**
+	 * Returns a sub-biomes for biome or null;
+	 * @param biome - parent {@link SmolderBiome};
+	 * @return {@link List} of {@link SmolderBiome} sub-biomes.
+	 */
+	public static List<SmolderBiome> getSubBiomes(SmolderBiome biome)
+	{
+		return SUBBIOMES.get(biome);
+	}
+	
+	private static void register(Identifier id, SmolderBiome biome)
+	{
+		if (!(biome instanceof SmolderWrappedBiome))
+			Registry.register(Registry.BIOME, id, biome);
+	}
+	
+	/**
+	 * Registers biome from another mod, used in callback.
+	 * @param i - raw ID;
+	 * @param id - {@link Identifier} of biome;
+	 * @param biome - {@link Biome} itself.
+	 */
+	public static void registerModBiome(int i, Identifier id, Biome biome)
+	{
+		registerBiome(biome);
+	}
+	
+	/**
+	 * Returns random {@link SmolderBiome}.
+	 * @param random - {@link Random};
+	 * @return weighted {@link SmolderBiome}.
+	 */
+	public static SmolderBiome getRandomBiome(Random random)
+	{
+		float w = random.nextFloat() * weight;
+		for (SmolderBiome biome: BIOMES)

```

honestly I don't really understand how this picks a random biome exactly but I'll take your word for it :P

I assume the default of nether wastes should never happen? I'd maybe throw an error in the log or something in addition.

---

Nether wastes have a default chance to be with other biomes, but if there is no any biomes (all weights is zero) - the nether wastes will be returned, and the Nether will be like in 1.12. The biome must never be null in output

---

Ok, that makes sense I guess